vars/promoteTarArtifacts.groovy

```diff
@@ -16,78 +16,94 @@ void call(Map args = [:]) {
-                String pluginNameWithExt = pluginSubPath.split('/')[1]
-                String pluginName = pluginNameWithExt.replace('-' + version + '.zip', '')
-                String pluginNameNoExt = pluginNameWithExt.replace('-' + version, '')
-                String pluginFullPath = ['plugins', pluginName, version].join('/')
-                s3Upload(
-                    bucket: "${ARTIFACT_PRODUCTION_BUCKET_NAME}",
-                    path: "releases/$pluginFullPath/",
-                    workingDir: "$WORKSPACE/artifacts/$artifactPath/builds/$filename/core-plugins/",
-                    includePathPattern: "**/${pluginName}*"
-                )
+        println("Signing Core/Bundle Artifacts")
+        String coreFullPath = ['core', filename, revision].join('/')
+        String bundleFullPath = ['bundle', filename, revision].join('/')
+        for (Closure action : fileActions) {
+            for (file in findFiles(glob: "**/${filename}-min-${revision}*.${distribution_local}*,**/${filename}-${revision}*.${distribution_local}*")) {
+                argsMap['artifactPath'] = "$WORKSPACE" + "/" + file.getPath()
+                action(argsMap)
             }
         }
 
-        s3Upload(
-            bucket: "${ARTIFACT_PRODUCTION_BUCKET_NAME}",
-            path: "releases/$coreFullPath/",
-            workingDir: "$WORKSPACE/artifacts/$artifactPath/builds/$filename/dist/",
-            includePathPattern: "**/${filename}-min-${version}*")
-
+        //////////// Uploading Artifacts
+        withAWS(role: "${ARTIFACT_PROMOTION_ROLE_NAME}", roleAccount: "${AWS_ACCOUNT_ARTIFACT}", duration: 900, roleSessionName: 'jenkins-session') {
+            // Core Plugins only needs to be published once through Tar, ignore other distributions
+            if(corePluginDirExists && distribution_local.equals('tar')) {
+                List<String> corePluginList = buildManifest.components.artifacts."core-plugins"[0]
+                for (String pluginSubPath : corePluginList) {
+                    String pluginSubFolder = pluginSubPath.split('/')[0]
+                    String pluginNameWithExt = pluginSubPath.split('/')[1]
+                    String pluginName = pluginNameWithExt.replace('-' + revision + '.zip', '')
+                    String pluginNameNoExt = pluginNameWithExt.replace('-' + revision, '')
+                    String pluginFullPath = ['plugins', pluginName, revision].join('/')
+                    s3Upload(
+                        bucket: "${ARTIFACT_PRODUCTION_BUCKET_NAME}",
+                        path: "releases/$pluginFullPath/",
+                        workingDir: "$prefixPath/$artifactPath/builds/$filename/core-plugins/",
+                        includePathPattern: "**/${pluginName}*"
+                    )
+                }
+            }
+            
+            // We will only publish min artifacts for Tar, ignore other distributions
+            if (distribution_local.equals('tar')) {
```

Comment:
This is a bad smell. Is there a way to extract the "special case of handling tar" out of the generic "promoteArtifacts"?

Reply:
They all have the same file name conversion except the distribution, hence this is the best usage.
We will only release tar min and that will not change for now.